### ISSUE TITLE
INC0032789 Sysdig agents disconnecting

### ISSUE THREAD

**Issue by sbarre-esit** [BCDevOps/developer-experience#327]
https://trello.com/c/VxaChI02/171-inc0032789-sysdig-agents-disconnecting

On Sunday I started getting a lot of flapping alerts for Host Down on the sysdig agents.

Sysdig case 00005175 - Agents unable to connect

Sysdig: 8051 - Metrics flapping or unavailable

**Comment by sbarre-esit** [comment]
Sysdig support has asked we set `go_k8s_user_events: false` in the configmap, and increase the CPU requests to 2.

Currently blocked by the RFC to set `kube-reserved` on all the nodes. We need to re-asses how full the cluster is after to understand if we can give another CPU core to Sysdig or not.

**Comment by wmhutchison** [comment]
Node reservation RFC completed successfully. 

The custom grafana page we use to track remaining APP node capacity didn't adjust accordingly, meaning we need to review and apply fixes to said graph first since it's likely going off original capacity instead of actual capacity. 

Once that is done and we confirm we're comfortable with the amount of resource loss involved, then a suitable RFC can be scheduled for implementing the vendor recommendation,

**Comment by sbarre-esit** [comment]
RFC194136 changed the go_k8s_user_events to false and increased the CPU limit to 4. Still experiencing issues and sent some more logs to Sysdig.

From: Sysdig Support [mailto:support@sysdig.com]
Sent: May 14, 2020 3:33 AM
Subject: Case: 00005175 - Agents unable to connect [ ref:_00Dj01pqKW._5003Z12F18U:ref ]

Thanks Steven for the fresh set of logs.

From the logs, I can see Sysdig agent process getting killed due to resource issue. I do understand having 4 cpu cores and 4 GB of memory is a little excessive. I have escalated the issue to our Agent development team to get their recommendation on next course of action to stabilize several of the agents connectivity.

I'll update again as soon as I have any feedback from our dev team.

**Comment by sbarre-esit** [comment]
Sysdig recommends we "reduce prometheus metrics scrape by agent by reducing the "- include" in the "process_filter" section of prometheus"

Shea is going to open a PR for https://github.com/BCDevOps/platform-services/blob/master/monitoring/sysdig/manifests/prod/cm-sysdig-agent-app.yaml to adjust after some investigating

**Comment by stewartshea** [comment]
Related to: https://github.com/BCDevOps/developer-experience/issues/491

**Comment by mitovskaol** [comment]
@stewartshea Can you please provide update on testing the fix from #491.

**Comment by stewartshea** [comment]
I'm still working on reducing  / refining the prometheus configuration in the lab. Hopefully this testing will be done by next week @mitovskaol

**Comment by sbarre-esit** [comment]
Applied updated configmaps from @stewartshea and upgraded to the latest agent version v10.1.1

Will check in the morning to see if agents are still crashing out.

**Comment by stewartshea** [comment]
Thanks @sbarre-esit ... Based on what I quickly saw they might still be crashing but we can confirm in the morning

**Comment by tosazuwa** [comment]
Waiting on vendor response.

**Comment by stewartshea** [comment]
Yeah, still troubleshooting with the vendor. The latest updates have not reconciled the issue.

**Comment by sbarre-esit** [comment]
No reply from Sysdig in a week. Followed up and CCd our new TAM.

**Comment by stewartshea** [comment]
Possibly figuring out how to remove other stats from collection that we aren't using today based on feedback from support.

**Comment by stewartshea** [comment]
Created PR https://github.com/BCDevOps/platform-services/pull/697 to disable additional metric checks to see if we can fully reduce the crashing. @sbarre-esit can you apply the configmap changes and reload the agents in prod?

**Comment by wmhutchison** [comment]
@stewartshea Requested configmap changes and agent refresh late yesterday afternoon. Not sure if it's still a problem since we do still have active ongoing OS patching taking place today and thus agent availability will still go up/down. Will revisit later when Tatsuya finishes his current round of OS patching.

**Comment by sbarre-esit** [comment]
Still seeing errors at 4am today, https://chat.pathfinder.gov.bc.ca/channel/platform-prod-alerts-sysdig?msg=fvPhtxD2RfpX7jDmz so likely not fixed.

**Comment by stewartshea** [comment]
I'm not even sure it's CPU related since those pods don't seem to have that type of spike. 

Lets follow the guidelines from the updated support ticket... Update the version and collect some logs again

**Comment by wmhutchison** [comment]
Per mention, Shea Stewart performed the request for additional logs which involved increasing log verbosity for three nodes, after which time the log verbosity was rolled back. Sysdig now has the log data dump for further analysis.

**Comment by wmhutchison** [comment]
Latest response from Sysdig clarified that additional log files that should have been created during the additional logging are not present. Will perform the original vendor request and re-upload after first confirming the new pods have the additional log files.

**Comment by wmhutchison** [comment]
Re-did verbose logging for hosts 185/186 and uploaded to sysdig. Confirmed they did receive the data and are currently going through it.

Sysdig support asked if there are still agent pods restarting and/or alerting for disconnects to back-end, confirmed this is so. Asked if they wanted more logs from those agents/nodes. Waiting for their response.

**Comment by wmhutchison** [comment]
No further response yet from Sysdig support regarding additional info required from us. 

Consensus during this week's meeting with Sysdig TAM was for them to follow up Sysdig Support and emphasize urgency of our need to resolve this ongoing stability issue.

**Comment by wmhutchison** [comment]
Pushed again this week for a status update from Sysdig support. They confirmed that they do not need any more data from us at this time, and have been able to replicate our issue in a controlled fashion in their environment, so relevant teams are now tasked to work on a fix to this issue, no ETA provided when that would be ready.

**Comment by wmhutchison** [comment]
Ball's still in vendor's court on this matter as they are working on a solution for what they have been able to replicate in their own test environment. Will push for a status update early next week.

**Comment by wmhutchison** [comment]
From: Sysdig Support [mailto:support@sysdig.com] 
Sent: July 29, 2020 6:00 AM
To: shea.stewart@arctiq.ca; twinkle.patel@sysdig.com; Baker, Tim <Tim.Baker@dxcas.com>; Hutchison, William <William.Hutchison@dxcas.com>; Barre, Steven <Steven.Barre@dxcas.com>
Cc: bcdevex@arctiq.ca; Baker, Tim <Tim.Baker@dxcas.com>; Hutchison, William <William.Hutchison@dxcas.com>; Morikawa, Tatsuya <Tatsuya.Morikawa@dxcas.com>; twinkle.patel@sysdig.com; shea.stewart@arctiq.ca
Subject: Case: 00005175 - Agents unable to connect [ ref:_00Dj01pqKW._5003Z12F18U:ref ]

Hello,
 
Sysdig agent version 10.3.0 has been released which has the fix for the agent restart issue seen in version 10.2.0. Please plan to update to this version at earliest.
 
Thanks,
Thomas


Thomas Ko

**Comment by wmhutchison** [comment]
Applying new agent in 3.11 LAB quickly, and assuming agent still works, will start doing this first in PROD for specific nodes with existing high restart counts, and if nothing breaks there, finish the rest.

**Comment by wmhutchison** [comment]
Results are somewhat positive. No new pod restarts are being seen, but we are still receiving the occasional report from Sysdig that an agent was disconnected. Will see if we can determine which agents on which nodes that is for, and then ask Sysdig support what they want us to do next.

**Comment by wmhutchison** [comment]
Ongoing back-and-forth with sysdig support still taking place for this - some new log requests were made by them and fulfilled by us. Currently waiting for them to finish the log analysis.

At present we're currently seeing about one agent disconnect per day.

**Comment by wmhutchison** [comment]
An additional request was made by Sysdig support for us to increase logging to full debug levels. Have done so, which has put a significant load on our EFK logging stack (500+GB/day). Have mitigated that by adding a custom rotation config for the sysdig project to only keep one days worth of logs so we don't fill up all available disk space for Elasticsearch.

The new challenge now that's been passed along to Sysdig support for a response on is regarding amount of log data stored by the agents. At present the logs local to the agent pods only keep 100MB worth, which amounts to a maximum of 10 minutes' worth of logs, which isn't nearly enough time for us to go to the pod to download the contents once an event is triggered.

One such event was in fact triggered by ociopf-p-183.dmz the evening of August 12th, but as stated, actual agent pod log retention wasn't nearly long enough for us to fetch the actual logs for that event in time. We also will have limits as to how much space we can reliably take up in overlay disk space as well.

Will see what Sysdig support returns back to us. In the meantime, there has been only a total of two recorded events in terms of downtime for the entire week (one was for a master node when we didn't have debug logging enabled for masters yet).

**Comment by wmhutchison** [comment]
Also will look into tweaking curator config to see if we can get away with 2 days' worth of retention for the sysdig project or not. If so, then it's possible we may be able to data-mine what the sysdig vendor support are looking for from Kibana. Will let on-call know about this change first so they are aware if they are notified regarding disk space concerns for Elasticsearch.

**Comment by wmhutchison** [comment]
Catching up on missed updates.

We were able to run for a while with full debug logging enabled, but then discovered that the sysdig pod log spam was crowding out all of the remaining log data on all the APP and MASTER nodes. Had to disable debug logging for a little while to fix regular logs being received by EFK logging.

Reviewed vendor settings once again. It asks us to set to debug level both the file logs as well as the console logs. The latter is what gets captured and injected into EFK Logging, while the file logs are what we actually fetch for Sysdig support. As such the console debug log level is completely un-necessary, so adjusted console logs back to normal levels while leaving file logs at debug.

Net result is we're still getting debug logging, but it's not over-loading EFK logging stack as a result. Updated the vendor on all of these findings since we did miss a few events while debug logging was disabled. Will update sysdig support when there's a new event to note.

**Comment by mitovskaol** [comment]
Monitor the disconnect events starting from Aug 31 for 2 weeks to confirm data capture for Sydig. In the week of Sep 14 review the number of disconnect events and decide on the appropriate amount of disconnects that is still acceptable for us.
If the number of disconnect events is considered acceptable, close the open ticket with Sysdig.

**Comment by wmhutchison** [comment]
Status of disconnect issue. Back-end noted some events, but they correlated to things not related to the general issue (monitor catching pods in the middle of re-deploying, etc).

Noted that Shea mentioned on devops-sysdig channel that he recently downgraded sysdig agent from "latest" to 10.3.1 specifically to deal with an issue involving users not seeing all workloads. Does not appear to have affected the current stability of our disconnect issue.

**Comment by wmhutchison** [comment]
We continue to monitor while the week of September 15th approaches which is when our final evaluation will take place as to whether or not we consider this issue resolved or not.

While there has been some alert notices regarding disconnects, they do correlate to work being done this week for PROD OS patching, which of course will disrupt all pods running on the node (sysdig agents do not evacuate when node is drained). Continuing to monitor while we correlate any new events with maintenance activities or otherwise take action to dump new log files for Sysdig support.

**Comment by wmhutchison** [comment]
There appears to be still the occasional disconnect events, but when Sysdig support look into it, all appear to be related to internal issues - error message states these are due to ERR_CONN_LIMIT events which they are looking into. This error came up last month and Sysdig supported incorrectly assumed it was a licensing issue which Shea Stewart quickly corrected them. 

Have uploaded a few more debug log dumps but am now stopping until Sysdig support states those are stilll needed/helpful for troubleshooting. The important thing to note is that these disconnect events are not the ones originally reported which was a client-side issue with the agent pods, but are now a back-end issue with the cloud-hosted SaaS platform for the service.

**Comment by wmhutchison** [comment]
Agent version 10.5.0 was identified by Sysdig support as a version we need to apply to address some of the issses involving master nodes remainng disconnected for extended periods of time. Confirmed with Shea Stewart regarding issues with this  (no issues).

Continuing to monitor, but starting to look pretty stable now. Will discuss with Shea Stewart and Sysdig TAM today in our meeting with Sysdig (monthly) about this, but it looks like we've achieved an acceptable level of stability, considering that the data involve is going offsite to the cloud, thus external connectivity issues need also to be factored into this. If frequency of disconnects is low and duration of disconnect is also low (several minutes), then we'll consider this officially Done.
